pallets/manta-pay/src/lib.rs

```diff
@@ -186,6 +190,17 @@ pub mod pallet {
         #[transactional]
         pub fn to_private(origin: OriginFor<T>, post: TransferPost) -> DispatchResultWithPostInfo {
             let origin = ensure_signed(origin)?;
+
+            let asset_id = post
+                .asset_id
+                .ok_or(Error::<T>::PublicUpdateInvalidAssetId)?;
+            for source in post.sources.iter() {
+                ensure!(
```

Comment:
I'm also thinking that a per-extrinsic limit is too simplistic.
The idea is to prevent large amounts per block, not per TX. While we could just make the limit small per-tx so it adds up to the intended limit per block, it'll make the transaction limit unnecesarily small, annoying normal users who might be the only one attempting a 5k value transaction in the current block, which we'd have to make fail.

IMHO, `pallet-tx-limit` should
- have a storage map `LimitUsedThisBlock<asset_id,Balance>`
- subscribe to `to_private` events posted to the chain, read them including transferred balance ( i don't think the latter gets sent with the Event currently )
- Increment the storage item
- Query storage on every `to_private` the way it's currently implemented but compare `storage item + tx amount < limit`
- Reset storage on every `on_initialize`

Reply:
Actually, It's difficult to using polkadot.js to filling blocks with many `to_private`, because `TransferPost` is kind of black box for user that it's generated by signer. The first challenge when I trying use polkadot.js is how do I fill `utxo`, If I use my previous tx information that already finalized, that's also wrong, because `utxo` can't be same. so I think it's fine for per-extrinsic limit.

Reply:
If I understand your concern correctly, it's about pre-computing the extrinsic data for multiple `to_private` txns in sequence without waiting for them to be included on-chain.
This whole issue can be sidestepped by splitting assets into multiple wallets with multiple signers and private ledgers.

The total number of TXNs that can be executed in parallel is limited by the max block weight however, so the transaction number per-time is still limited.